compiler/quilt/nodes.py

```diff
@@ -93,6 +99,47 @@ def _add_group(self, groupname):
         child = GroupNode(self._package, core.GroupNode({}))
         setattr(self, groupname, child)
 
+    @classmethod
+    def __iteritems(cls, node, base_path, recursive=False):
```

Comment:
Why are you adding these helper methods? It's not obvious (to me) why they're needed and how they're used. If possible, we should try to reuse and possibly extend exiting helpers and APIs instead of duplicating functionality.

Reply:
Clarity in usage.

* Without `DataNode._datafile`:
  * In order to determine what type of data their `DataNode` has, the user must do one of the following:
    * Call `node._data()`, potentially loading a very large amount of data
    * Know that `quilt.nodes.DataNode._node` is either a `quilt.tools.core.FileNode` or a `quilt.tools.core.TableNode`, and check for what they want
    * The resultant code is less succinct and less clear.
* With `DataNode._datafile`:
 * The user needs to know that if the `DataNode` references file data, it's available as `DataNode._datafile`.
 * The resultant code is a one-liner that is very clear.

```
## Minimal Example Without DataNode._datafile:
from quilt.tools.core import FileNode
from quilt.data.user import mydata

files = {}
for key in mydata._data_keys():
    node = getattr(mydata, key)
    if isinstance(node._node, FileNode):
        files[key] = node._data()

# or, as a comprehension, this *is* fairly compact -- however, it still requires internal knowledge of quilt,
# and isn't easy on the eyes.
files = {key: getattr(mydata, key)._data() for key in mydata._data_keys()
            if isinstance(getattr(mydata, key)._node, FileNode)}
# Not too bad, but it's not exactly clear, either.

## Minimal Example With DataNode._datafile:
from quilt.data.user import mydata

files = {}
for key, node in mydata._items():
    if node._datafile:
        files[key] = node._datafile

# Or, as a comprehension, clear and concise:
files = {key: node._datafile for key, node in mydata._items() if node._datafile}
```

As to `DataNode._iteritems` and `DataNode._iterpaths`:
Mine does recursion.  I'll just add it as an optional parameter to `_items()` and `_keys` instead.

Reply:
They're recursive, and I was avoiding touching the existing `_items` and `_keys`.  The recursivity is used in export, and is also a good user-facing feature.

I've shifted the code over to `_items()` and `_keys()`.

Reply:
Let's try to consolidate helper functions like these in the Package class so we don't end up with two full implementations of tree classes that are subtly different.